// Copyright 2010-2013 Google
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
syntax = "proto2";

package operations_research.sat;

// Contains the definitions for all the sat algorithm parameters and their
// default values.
message SatParameters {
  // Tie breaking for the variable activity. If two variables have the same
  // activity, then the one with higher weight will be selected first for
  // branching.
  enum VariableWeight {
    // Zero by default but can be changed with SetAssignmentPreference().
    // This stays the same between Solve() calls.
    DEFAULT_WEIGHT = 0;
    // Both value below are recomputed at the beginning of each Solve() call.
    RANDOM_WEIGHT = 1;
    STATIC_SCALED_USAGE_WEIGHT = 2;
  }
  optional VariableWeight variable_weight = 1 [default = DEFAULT_WEIGHT];

  // Specifies how branching on variable should be done.
  // Note that this is overridden by calls to SetAssignmentPreference().
  enum VariableBranching {
    // Always branch on the variable being true.
    FIXED_POSITIVE = 0;
    // Always branch on the variable being false.
    FIXED_NEGATIVE = 1;
    // The sign of a variable is true if it appears in more clauses in a
    // positive literal than in clauses in a negative literal. With that
    // definition, branching will follow the sign of the variable.
    SIGN = 2;
    // Branching will follow the reverse of the sign.
    REVERSE_SIGN = 3;
    // The polarity of the variable is the last value it was assigned
    // to, or its sign if it was never assigned.
    // Note this is also referenced as 'literal phase saving' in some papers.
    // For details see 'A Lightweight Component Caching Scheme for
    // Satisfiability Solvers' K. Pipatsrisawat and A.Darwiche, In 10th
    // International Conference on Theory and Applications of Satisfiability
    // Testing, 2007.
    POLARITY = 4;
    // Branching will follow the reverse of the polarity.
    REVERSE_POLARITY = 5;
  }
  optional VariableBranching variable_branching = 2 [default = POLARITY];

  // Specifies how literals should be initially sorted in a clause.
  enum LiteralOrdering {
    // Do nothing and keep literals in their current order.
    LITERAL_IN_ORDER = 0;
    // Sort the literals by increasing order of their variable appearance in the
    // problem.
    VAR_MIN_USAGE = 1;
    // Sort the literals by decreasing order of their variable appearance in the
    // problem.
    VAR_MAX_USAGE = 2;
  }
  optional LiteralOrdering literal_ordering = 3 [default = LITERAL_IN_ORDER];

  // Do we try to minimize conflicts (greedily) when creating them.
  enum ConflictMinimizationAlgorithm {
    NONE = 0;
    SIMPLE = 1;
    RECURSIVE = 2;
    EXPERIMENTAL = 3;
  }
  optional ConflictMinimizationAlgorithm minimization_algorithm = 4
      [default = RECURSIVE];

  // Each time the learned clause database is cleaned, we want the target size
  // of the next cleaning to be equals to the current database after it has just
  // been cleaned plus this parameter.
  optional double clause_cleanup_increment = 11 [default = 1500];

  // Deletes this ratio of clauses during each cleanup.
  optional double clause_cleanup_ratio = 13 [default = 0.5];

  // Variable activity parameters.
  //
  // Each time a conflict is found, the activities of some variables are
  // increased by one. Then, the activity of all variables are multiplied by
  // variable_activity_decay.
  //
  // To implement this efficiently, the activity of all the variables is not
  // decayed at each conflict. Instead, the activity increment is multiplied by
  // 1 / decay. When an activity reach max_variable_activity_value, all the
  // activity are multiplied by 1 / max_variable_activity_value.
  optional double variable_activity_decay = 15 [default = 0.8];
  optional double max_variable_activity_value = 16 [default = 1e100];

  // The activity starts at 0.8 and increment by 0.01 every 5000 conflicts until
  // 0.95. This "hack" seems to work well and comes from:
  //
  // Glucose 2.3 in the SAT 2013 Competition - SAT Competition 2013
  // http://edacc4.informatik.uni-ulm.de/SC13/solver-description-download/136
  optional double glucose_max_decay = 22 [default = 0.95];
  optional double glucose_decay_increment = 23 [default = 0.01];
  optional double glucose_decay_increment_period = 24 [default = 5000];

  // Clause activity parameters (same effect as the one on the variables).
  optional double clause_activity_decay = 17 [default = 0.999];
  optional double max_clause_activity_value = 18 [default = 1e20];

  // Whether or not we take the LBD (literal blocks distance) into account
  // during the conflict cleaning phase. If true, then clauses with an LBD of
  // 2 will not be deleted and clauses will be ordered by increasing LBD first
  // with a tie breaking given by decreasing activity.
  optional bool use_lbd = 20 [default = true];

  // To try to reward good variables, Glucose bumps again the variable from the
  // last decision level and with a learned reason of smaller LBD than the 1 UIP
  // conflict. This needs use_lbd() to be true.
  optional bool use_glucose_bump_again_strategy = 21 [ default = false];

  // Frequecy of periodic restart if > 0. A negative value indicates
  // no restart.
  optional int32 restart_period = 30 [default = 100];

  // At the beginning of each solve, the random number generator used in some
  // part of the solver is reinitialized to this seed. If you change the random
  // seed, the solver may make different choices during the solving process.
  //
  // For some problems, the running time may vary a lot depending on small
  // change in the solving algorithm. Running the solver with different seeds
  // enables to have more robust benchmarks when evaluating new features.
  optional int32 random_seed = 31 [default = 1];

  // A number between 0 and 1 that indicates the proportion of branching
  // variables that are selected randomly instead of choosing the first variable
  // from the given variable_ordering strategy.
  optional double random_branches_ratio = 32 [default = 0];

  // If true, the binary clauses are treated separately from the others. This
  // should be faster and uses less memory. However it changes the propagation
  // order.
  optional bool treat_binary_clauses_separately = 33 [default = true];

  // Whether to expoit the binary clause to minimize learned clauses further.
  // This will have an effect only if treat_binary_clauses_separately is true.
  enum BinaryMinizationAlgorithm {
    NO_BINARY_MINIMIZATION = 0;
    BINARY_MINIMIZATION_FIRST = 1;
    BINARY_MINIMIZATION_WITH_REACHABILITY = 2;
    EXPERIMENTAL_BINARY_MINIMIZATION = 3;
  }
  optional BinaryMinizationAlgorithm binary_minimization_algorithm = 34
      [default = BINARY_MINIMIZATION_FIRST];

  // For an optimization problem, whether we follow some hints in order to find
  // a better first solution. For a variable with hint, the solver will always
  // try to follow the hint. It will revert to the variable_branching default
  // otherwise.
  optional bool use_optimization_hints = 35 [default = true];

  // Maximum time allowed in seconds to solve a problem.
  // The counter will starts as soon as Solve() is called.
  optional double max_time_in_seconds = 36 [default = inf];

  // Maximum number of conflicts allowed to solve a problem.
  optional int64 max_number_of_conflicts = 37
      [default = 0x7FFFFFFFFFFFFFFF];  // kint64max

  // Maximum memory allowed for the whole thread containing the solver. The
  // solver will abort as soon as it detects that this limit is crossed. As a
  // result, this limit is approximative, but usually the solver will not go too
  // much over.
  optional int64 max_memory_in_mb = 40 [default = 4096];

  // Whether the solver should log the search progress to LOG(INFO).
  optional bool log_search_progress = 41 [default = false];

  // Indicates if the solver maintain in memory the information needed to
  // generate an UNSAT core if the problem is unsat or to generate a full
  // resolution proof. This can potentially use a lot of memory and may slow
  // down the solver a bit.
  optional bool unsat_proof = 42 [default = false];
}
